SPO-13135 Анимация

ingvilow opened the pull request.

ingvilow committed:
SPO 13135 Анимация

ingvilow committed:
no-task cross fade animation

maxim-sysoev commented on lib/animations/animation_switcher_demo/animated_switcher.dart:
```diff
@@ -0,0 +1,50 @@
+import 'package:elementary/elementary.dart';
```
странное расположение фалов, почему этот файл находится  отдельной папке от wm and model?

ingvilow replied:
у меня тут была гениальная идея использовать один wm на все, но что-то пошло не так. Не сразу до меня дошло как это сделать и не плодить wm

maxim-sysoev replied:
тогда стоит перенести к остальным файлам, чтобы не нарушать структуру

maxim-sysoev commented on lib/animations/animated_switcher_wm.dart:
```diff
@@ -0,0 +1,52 @@
+  final StateNotifier<bool> _isRotates = StateNotifier();
+
+  @override
+  bool get isRotate => _isRotate;
```
а почему is rotate?  вроде как используется другая анимация

maxim-sysoev commented on lib/animations/cross_fade/cross_fade_model.dart:
```diff
@@ -0,0 +1,3 @@
+import 'package:elementary/elementary.dart';
+
+class CrossFadeAnimationModel extends ElementaryModel{}
```
форматирование не сработало, нет пустой строки в конце файла

maxim-sysoev commented on lib/animations/cross_fade/cross_fade_wm.dart:
```diff
@@ -0,0 +1,43 @@
+  Duration get duration => _duration;
+
+  @override
+  ListenableState<bool> get isRotates => _isRotates;
```
почему снова is rotate?

maxim-sysoev commented on lib/animations/cross_fade/cross_fade_animation.dart:
```diff
@@ -0,0 +1,49 @@
+                    height: 200,
+                    color: Colors.blueGrey,
+                  ),
+                  crossFadeState: wm.isRotate
```
не обязательно тут указывать значение из wm, лучше использовать value из билдера
если указать дженерик тип у `StateNotifierBuilder<bool>`, то он будет нормально работать

ingvilow replied:
`StateNotifierBuilder<bool>` - в этом случае падает в эту ошибку: `Null check operator used on a null value` при попытке использовать value. Я так поняла (если я ничего не путаю, потому что помню, что Артем говорил о такой особенности), что StateNotifierBuilder делает null-абельным по умолчанию

arefimenko replied:
там проблема в геттере у `ListenableState`, поэтому "стэйты" из elementary я бы лично использовал только в случае если поле реально будет nullable, для других случаев использовать `ValueNotifier` в wm + `ValueListenableBuilder` на ui

ingvilow committed:
no-task анимация смены цвета иконки

ingvilow committed:
SPO-13135 анимация

ingvilow committed:
SPO-13135 animation

maxim-sysoev commented on lib/animations/animated_switcher_wm.dart:
```diff
@@ -0,0 +1,52 @@
+class AnimateWM extends WidgetModel<AnimatedSwitcherScreen, AnimationModel>
+    implements IAnimate {
+  //утснавлиаем длительноть анимации
+  final _duration = const Duration(milliseconds: 600);
```
переменную можно сделать const вне класса, это нормальная практика

ingvilow committed:
no-task ScaleTransition анимация

ingvilow committed:
no-task SizeTransition, смена цвета иконки, смена самой иконки

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_screen.dart:
```diff
@@ -0,0 +1,67 @@
+            ),
+          ),
+          ExpansionTile(
+            // onExpansionChanged:,
```
лишнее

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_screen.dart:
```diff
@@ -0,0 +1,67 @@
+                },
+              ),
+            ],
+            //onExpansionChanged: true,
```
лишнее

ingvilow committed:
no-task поправила замечания из пр по смене со StateNotifierBuilder на ValueListenableBuilder. В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_wm.dart:
```diff
@@ -0,0 +1,83 @@
+    curve: Curves.fastOutSlowIn,
+  );
+
+  @override
+  AnimationController? get controller => _controller;
+
+  @override
+  Animation<double> get scaleAnimation => _animation;
```
это можно сделать без геттера, экономим строки и пишем меньше... одни плюсы

ingvilow replied:
+

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_wm.dart:
```diff
@@ -0,0 +1,83 @@
+  );
+
+  @override
+  AnimationController? get controller => _controller;
```
контроллер анимации скорее всего не нужен на слое ui, там только анимация вроде как нужна

ingvilow replied:
+

ingvilow committed:
no-task поправила замечания по оформлению документации.

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_screen.dart:
```diff
@@ -0,0 +1,65 @@
+                    child: ListTile(
+                      title: const Text('important data 1'),
+                      trailing: IconButton(
+                        icon: value
+                            ? const Icon(
+                                Icons.favorite_border,
+                                color: Colors.purple,
+                              )
+                            : const Icon(
+                                Icons.favorite,
+                                color: Colors.pink,
```
не похоже на анимацию плавную

ingvilow replied:
там в ноушене написано просто реализовать смену цвета иконки, без всяких эффектов, вот я и сделала так)

ingvilow replied:
+

ingvilow committed:
no-task сделала плавную анимацию для иконки. Использовала AnimatedSwitcher

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_wm.dart:
```diff
@@ -0,0 +1,75 @@
+  @override
+  ValueListenable<bool> get isExpands => _isExpands;
+
+  @override
+  Animation<double> get sizeTransitionAnimation => _sizeTransitionAnimation;
+
+  @override
+  AnimationController? get sizeTransitionController =>
+      _sizeTransitionController;
```
можно без геттера

maxim-sysoev replied:
контроллер не нужен на ui слое

ingvilow replied:
там в ui в ScaleTransition надо будет передать

```
ScaleTransition(
             scale: wm.sizeTransitionAnimation,
```

ingvilow replied:
он без него ругается, это обязательный аргумент

maxim-sysoev replied:
да, я к тому что можно сразу сделать `override` таким образом
```dart
@override
late final Animation<double> sizeTransitionAnimation = CurvedAnimation(
  parent: _controller,
  curve: Curves.linear,
);
```

а вот контроллер анимации можно оставить приватным и не объявлять для него геттер в интерфейсе, т.к. он не используется в ui

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_wm.dart:
```diff
@@ -0,0 +1,75 @@
+
+  final ValueNotifier<bool> _isExpands = ValueNotifier(false);
+
+  late final AnimationController _controller = AnimationController(
+    duration: const Duration(seconds: 15),
+    vsync: this,
+  )..repeat(reverse: true);
+  late final Animation<double> _animation = CurvedAnimation(
+    parent: _controller,
+    curve: Curves.fastOutSlowIn,
+  );
```
используется где-то?

ingvilow replied:
ну по сути, до того, как я убрала тот контроллер с анимацией, то эта часть кода отвечала за то, чтобы анимации не выполнялись одновременно. Сейчас это либо убирать, потому у меня уже есть сверху кусок кода, который управляет анимацией и тогда они обе на экране будут выполнятся одновременно, потому что в ScaleTransition(
             scale: wm.sizeTransitionAnimation я передаю обязательный аргумент для этого типа анимации, либо вернуть назад контроллеры, которые были прилеплены именно к этому куску кода и тогда анимации будут выполнятся не одновременно

maxim-sysoev replied:
можешь оставить один контроллер, и одну анимацию, или сделать 2 анимации от одного контроллера, или 2 контроллера и две анимации, но главное чтобы все создаваемые переменные использовались
оставлять неиспользуемые переменные в коде не очень хорошая практика, легко запутаться и поддерживать не кайф. поэтому постарайся следить чтобы не было таких моментов

ingvilow committed:
no-task сделала Tween-анимацию для контейнера, использовала AnimatedBuilder.

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

ingvilow committed:
no-task поправила форматирование

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

maxim-sysoev commented on lib/animations/scale_transition/scale_transition_wm.dart:
```diff
@@ -0,0 +1,72 @@
+  @override
+  ValueListenable<bool> get isExpands => _isExpands;
+
+  @override
+  Animation<double> get sizeTransitionAnimation => _sizeTransitionAnimation;
+
+  @override
+  Animation<Size> get myAnimationTween => _myAnimation;
```
можно сделать без геттеров, сразу переопределить в публичные переменные

ingvilow committed:
no-task поправила из пр-а замечания по контроллеру

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9

ingvilow committed:
no-task сделала @override у Tween-анимации

 В этом коммите изменения по эти задачам из ноушен:

https://www.notion.so/AnimatedSwitcher-da51ef44e9364cb0b916828724384045

https://www.notion.so/SizeTransition-e7f43f243f16491d8f582b98516fe6ca

https://www.notion.so/AnimatedCrossFade-faf14e8866d14dcbb89616ad497b08bc

https://www.notion.so/FadeTransition-ScaleTransition-00cbff065e85465f80960fd91ddf01ae

https://www.notion.so/23fedd1b1938408dbcafa47e456247ae

https://www.notion.so/c0745a245f554e40a39f192b9df2a9a9
